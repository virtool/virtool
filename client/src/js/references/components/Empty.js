import React from "react";
import { connect } from "react-redux";
import { Alert, Button } from "../../base";
import { clearError } from "../../errors/actions";
import { getTargetChange } from "../../utils/utils";
import { emptyReference } from "../actions";
import { DataTypeSelection } from "./DataTypeSelection";
import { ReferenceForm } from "./Form";

const getInitialState = () => ({
    name: "",
    description: "",
    dataType: "genome",
    organism: "",
    errorName: "",
    errorDataType: "",
    mode: "empty"
});

export class EmptyReference extends React.Component {
    constructor(props) {
        super(props);
        this.state = getInitialState();
    }

    handleChange = e => {
        const { name, value, error } = getTargetChange(e.target);

        this.setState({
            [name]: value,
            [error]: ""
        });
    };

    handleChangeDataType = dataType => {
        this.setState({ dataType });
    };

    handleSubmit = e => {
        e.preventDefault();

        if (!this.state.name.length) {
            this.setState({ errorName: "Required Field" });
        }

        if (!this.state.dataType.length) {
            this.setState({ errorDataType: "Required Field" });
        }

        if (this.state.name.length && this.state.dataType.length) {
            this.props.onSubmit(this.state.name, this.state.description, this.state.dataType, this.state.organism);
        }
    };

    render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <Alert>
                    <strong>Create an empty reference.</strong>
                </Alert>
                <ReferenceForm
                    description={this.state.description}
                    errorFile={this.state.errorFile}
                    errorSelect={this.state.errorSelect}
                    errorName={this.state.errorName}
                    name={this.state.name}
                    mode={this.state.mode}
                    organism={this.state.organism}
                    onChange={this.handleChange}
                />
                <DataTypeSelection onSelect={this.handleChangeDataType} dataType={this.state.dataType} />
                <Button type="submit" icon="save" color="blue">
                    Save
                </Button>
            </form>
        );
    }
}

export const mapDispatchToProps = dispatch => ({
    onSubmit: (name, description, dataType, organism) => {
        dispatch(emptyReference(name, description, dataType, organism));
    },

    onClearError: error => {
        dispatch(clearError(error));
    }
});

export default connect(null, mapDispatchToProps)(EmptyReference);
