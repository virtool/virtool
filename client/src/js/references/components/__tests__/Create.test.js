import { EmptyReference, mapDispatchToProps } from "../Empty";

describe("<EmptyReference />", () => {
    const props = {
        onSubmit: jest.fn()
    };

    let state;

    const e = {
        preventDefault: jest.fn(),
        target: {
            name: "name",
            value: "foo",
            error: "error"
        }
    };

    beforeEach(() => {
        state = {
            name: "",
            description: "",
            dataType: "genome",
            organism: "",
            errorName: "",
            errorDataType: "",
            mode: "empty"
        };
    });

    it("should render", () => {
        const wrapper = shallow(<EmptyReference {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
    it("handleChange should update name and error", () => {
        const wrapper = shallow(<EmptyReference {...props} />);
        wrapper.find("ReferenceForm").simulate("change", e);
        expect(wrapper.state()).toEqual({
            ...state,
            name: "foo"
        });
    });

    it("handleSubmit() should update errorName in state when [this.state.name.length = 0]", () => {
        const wrapper = shallow(<EmptyReference {...props} />);
        wrapper.find("form").simulate("submit", e);
        wrapper.setState({ errorDataType: "foo" });
        expect(wrapper.state()).toEqual({
            ...state,
            errorDataType: "foo",
            errorName: "Required Field"
        });
    });

    it("handleSubmit() should update errorName in state when [this.state.dataType.length = 0]", () => {
        const wrapper = shallow(<EmptyReference {...props} />);
        wrapper.setState({
            ...state,
            name: "foo",
            dataType: ""
        });
        wrapper.find("form").simulate("submit", e);
        expect(wrapper.state()).toEqual({
            ...state,
            name: "foo",
            dataType: "",
            errorDataType: "Required Field"
        });
    });
    it("handleSubmit should call this.props.onSubmit when [this.state.name.length!=0] and [this.state.dataType.length!=0]", () => {
        const wrapper = shallow(<EmptyReference {...props} />);
        wrapper.setState({
            ...state,
            name: "foo",
            description: "bar",
            organism: "baz"
        });
        wrapper.find("form").simulate("submit", e);
        expect(props.onSubmit).toHaveBeenCalledWith("foo", "bar", "genome", "baz");
    });
});

describe("mapDispatchToProps()", () => {
    const dispatch = jest.fn();
    const props = mapDispatchToProps(dispatch);
    it("should return onSubmit in props", () => {
        props.onSubmit("foo", "bar", "fee", "baz");
        expect(dispatch).toHaveBeenCalledWith({
            name: "foo",
            description: "bar",
            dataType: "fee",
            organism: "baz",
            type: "EMPTY_REFERENCE_REQUESTED"
        });
    });
    it("should return onClearError in props", () => {
        props.onClearError("foo");
        expect(dispatch).toHaveBeenCalledWith({
            error: "foo",
            type: "CLEAR_ERROR"
        });
    });
});
