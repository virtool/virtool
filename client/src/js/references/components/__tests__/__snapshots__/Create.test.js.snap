// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmptyReference /> should render 1`] = `
<form
  onSubmit={[Function]}
>
  <Alert>
    <strong>
      Create an empty reference.
    </strong>
  </Alert>
  <ReferenceForm
    description=""
    errorName=""
    mode="empty"
    name=""
    onChange={[Function]}
    organism=""
  />
  <DataTypeSelection
    dataType="genome"
    onSelect={[Function]}
  />
  <Button
    color="blue"
    disabled={false}
    icon="save"
    tipPlacement="top"
    type="submit"
  >
    Save
  </Button>
</form>
`;
